src/aio.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (C) 2019 Alibaba Cloud Computing. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Safe wrapper for Linux AIO API.
+
+#![allow(non_camel_case_types)]
+
+/* automatically generated by rust-bindgen and then manually edited */
+
+use std::io::{Error, Result};
+use std::os::raw::{c_int, c_long, c_uint, c_ulong};
+use std::ptr::null_mut;
+
+type __s16 = ::std::os::raw::c_short;
+type __u16 = ::std::os::raw::c_ushort;
+type __u32 = ::std::os::raw::c_uint;
+type __s64 = ::std::os::raw::c_longlong;
+type __u64 = ::std::os::raw::c_ulonglong;
+
+pub const IOCB_CMD_PREAD: u32 = 0;
+pub const IOCB_CMD_PWRITE: u32 = 1;
+pub const IOCB_CMD_FSYNC: u32 = 2;
+pub const IOCB_CMD_FDSYNC: u32 = 3;
+pub const IOCB_CMD_NOOP: u32 = 6;
+pub const IOCB_CMD_PREADV: u32 = 7;
+pub const IOCB_CMD_PWRITEV: u32 = 8;
+pub const IOCB_FLAG_RESFD: u32 = 1;
+
+/// Limit maximum number of concurrent requests.
+pub const MAX_REQUESTS: usize = 0x10000;
+
+/// Rust version of iocb
+#[repr(C)]
+#[derive(Debug, Default, Copy, Clone)]
+pub struct IoControlBlock {
+    pub aio_data: __u64,
+    pub aio_key: __u32,
+    pub aio_reserved1: __u32,
+    pub aio_lio_opcode: __u16,
+    pub aio_reqprio: __s16,
+    pub aio_fildes: __u32,
+    pub aio_buf: __u64,
+    pub aio_nbytes: __u64,
+    pub aio_offset: __s64,
+    pub aio_reserved2: __u64,
+    pub aio_flags: __u32,
+    pub aio_resfd: __u32,
+}
+
+/// Rust version of io_event_t.
+#[repr(C)]
+#[derive(Debug, Default, Copy, Clone)]
+pub struct IoEvent {
+    pub data: __u64,
+    pub obj: __u64,
+    pub res: __s64,
+    pub res2: __s64,
+}
+
+/// Newtype for aio_context_t.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct IoContext(::std::os::raw::c_ulong);
```

Comment:
I don't know in detail how does the IoContext work, so it might be a dumb question but: is the inner value private on purpose?

Reply:
Yes, it's private on purpose, so it could only be accessed through exported public interfaces.